firedrake/mesh.py

```diff
@@ -1537,6 +1686,175 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
+        raise NotImplementedError("Extruded meshes are not supported")
+
+    if gdim != tdim:
+        raise NotImplementedError("Immersed manifold meshes are not supported")
+
+    if mesh.coordinates.function_space().ufl_element().degree() > 1:
+        raise NotImplementedError("Only straight edged meshes are supported")
+
+    if hasattr(mesh, "_made_from_coordinates") and mesh._made_from_coordinates:
+        raise NotImplementedError("Meshes made from coordinate fields are not yet supported")
+
+    if pdim != tdim:
+        raise ValueError(f"Mesh topological dimension {tdim} must match point list dimension {pdim}")
+
+    swarm = _pic_swarm_in_plex(mesh.topology._plex, vertexcoords, fields=[("parentcellnum", 1, IntType)])
+
+    dmswarm.label_pic_parent_cell_nums(swarm, mesh)
+
+    # Topology
+    topology = VertexOnlyMeshTopology(swarm, mesh.topology, name="swarmmesh", reorder=False)
+
+    # Geometry
+    tcell = topology.ufl_cell()
+    cell = tcell.reconstruct(geometric_dimension=gdim)
+    element = ufl.VectorElement("DG", cell, 0)
+    # Create mesh object
+    vmesh = MeshGeometry.__new__(MeshGeometry, element)
+    vmesh._topology = topology
+    vmesh._parent_mesh = mesh
+
+    def callback(self):
+        """Finish initialisation."""
+        del self._callback
+        # Finish the initialisation of mesh topology
+        self.topology.init()
+
+        coordinates_fs = functionspace.VectorFunctionSpace(self.topology, "DG", 0,
+                                                           dim=gdim)
+
+        coordinates_data = dmswarm.reordered_coords(swarm, coordinates_fs.dm.getDefaultSection(),
+                                                    (self.num_vertices(), gdim))
+
+        coordinates = function.CoordinatelessFunction(coordinates_fs,
+                                                      val=coordinates_data,
+                                                      name="Coordinates")
+
+        self.__init__(coordinates)
+
+    vmesh._callback = callback
+    return vmesh
+
+
+def _pic_swarm_in_plex(plex, coords, **kwargs):
+    """
+    Create a Particle In Cell (PIC) DMSwarm, immersed in a DMPlex
+    at given point coordinates.
+
+    This should only by used for dmplexes associated with meshes with
+    straight edges. If not, the particles may be placed in the wrong
+    cells.
+
+    :arg plex: the DMPlex within with the DMSwarm should be
+        immersed.
+    :arg coords: an `ndarray` of (npoints, coordsdim) shape.
+    :kwarg fields: An optional list of named data which can be stored
+        for each point in the DMSwarm. The format should be
+        `[(fieldname1, blocksize1, dtype1),
+          ...,
+         (fieldnameN, blocksizeN, dtypeN)]`.
+        For example, the swarm coordinates themselves are stored in a
+        field named `DMSwarmPIC_coor` which, were in not created
+        automatically, would be initialised with
+        `fields = [("DMSwarmPIC_coor", coordsdim, ScalarType)]`.
```

Comment:
RST is not markdown (boo) so I think you need double `

Reply:
tvm, changed